File: raytrace/src/shapes.rs
use crate::matrix::{Matrix, identity};
use crate::materials::{Material, material};
use crate::tuple::{Tuple, vector, point};
use crate::rays::Ray;
use crate::intersections::Intersection;
use crate::shape_index::{ShapeIndex, shape_index};
use core::fmt::Debug;

#[derive(Debug,PartialEq,Clone)]
pub struct Base {
    transform: Matrix,
    material: Material,
    i: ShapeIndex,
}

impl Shape for Base {
    fn get_transform(&self) -> &Matrix {
        &self.transform
    }

    fn set_transform(&mut self, t: Matrix){
        self.transform = t
    }

    fn get_material(&self) -> &Material {
        &self.material
    }

    fn set_material(&mut self, m: Material){
        self.material = m
    }

    fn local_normal_at(&self, object_point: Tuple) -> Tuple {
        vector(object_point.x, object_point.y, object_point.z)
    }

    fn intersect(&self, _r: Ray) -> Vec<Intersection> {
        vec!()
    }

    fn get_index(&self) -> usize {
        self.i.get_index()
    }

    fn set_index(&mut self, index: usize){
        self.i.set_index(index);
    }

    fn get_parent(&self) -> Option<usize> {
        self.i.parent
    }

    fn set_parent(&mut self, parent_index: usize){
        self.i.parent = Some(parent_index);
    }
}

pub trait Shape {
    fn get_transform(&self) -> &Matrix;
    fn set_transform(&mut self, t: Matrix);
    fn get_material(&self) -> &Material;
    fn set_material(&mut self, m: Material);
    
    fn normal_at(&self, world_point: Tuple) -> Tuple {
        let object_point = self.get_transform().inverse().multup( &world_point );
        let local_normal = self.local_normal_at( object_point );
        let mut world_normal = self.get_transform()
                                   .inverse()
                                   .transpose()
                                   .multup( &local_normal );
        world_normal.w = 0.0;
        world_normal.normal()
    }
    fn local_normal_at(&self, object_point: Tuple) -> Tuple;
    fn intersect(&self, r: Ray) -> Vec<Intersection>; 
    fn get_index(&self) -> usize;
    fn set_index(&mut self, index: usize);
    fn get_parent(&self) -> Option<usize>;
    fn set_parent(&mut self, parent_index: usize);

    // text implements this as a mutable field on Shape,
    // but this causes mutability contagion across the entire
    // project - trying this out as a query method instead
    // will probably want to figure out a better name here
    fn saved_ray(&self, r: Ray) -> Ray {
        r.transform( self.get_transform().inverse() )
    }

    fn world_to_object(&self, _p: Tuple) -> Tuple {
        //match self.get_parent() {
            //Some(parent) => parent.world_to_object(p),
            //None         => self.get_transform().inverse().multup(&p)
        //}
        point(0.0, 0.0, -1.0)
    }
}

impl Debug for dyn Shape {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "Shape {}", self.get_index())
    }
}

impl PartialEq for dyn Shape {
    fn eq(&self, _: &(dyn Shape + 'static)) -> bool {
        true   // need to research right way to compare trait objects
    }
}

pub fn shape() -> Base {
    Base {
        transform: identity(),
        material: material(),
        i: shape_index(),
    }
}

#[cfg(test)]
mod tests {
    use crate::matrix::identity;
    use crate::shapes::{Shape, shape};
    use crate::transform::{scaling, translation, rotation_y, rotation_z};
    use crate::materials::material;
    use crate::rays::ray;
    use crate::tuple::{point, vector};
    use crate::groups::group;
    use crate::spheres::sphere;
    use std::f64::consts::{PI, SQRT_2};

    #[test]
    fn shape_default_transformation(){
        let s = shape();
        assert!( s.get_transform().equals( identity() ));
    }

    #[test]
    fn assigning_a_transformation(){
        let mut s = shape();
        let t = translation( 2.0, 3.0, 4.0 );
        s.set_transform( t );
        assert!( s.get_transform().equals( translation( 2.0, 3.0, 4.0 ) ));
    }

    #[test]
    fn shape_default_material(){
        let s = shape();
        assert!( s.get_material().equals( material() ));
    }

    #[test]
    fn assigning_a_material(){
        let mut s = shape();
        let mut m = material();
        m.ambient = 1.0;
        s.set_material( m.clone() );
        assert!( !s.get_material().equals( material() ));
        assert!( s.get_material().equals( m ));
    }

    #[test]
    fn intersecting_scaled_shape_with_ray(){
        let r = ray( point(0.0, 0.0, -5.0), vector(0.0, 0.0, 1.0), 0 );
        let mut s = shape();
        s.set_transform( scaling(2.0, 2.0, 2.0) );

        let local_ray = s.saved_ray(r);
        
        assert!( local_ray.origin.equals( point(0.0, 0.0, -2.5) ));
        assert!( local_ray.direction.equals( vector(0.0, 0.0, 0.5) ));
    }

    #[test]
    fn intersecting_translated_shape_with_ray(){
        let r = ray( point(0.0, 0.0, -5.0), vector(0.0, 0.0, 1.0), 0 );
        let mut s = shape();
        s.set_transform( translation(5.0, 0.0, 0.0) );

        let local_ray = s.saved_ray(r);

        assert!( local_ray.origin.equals( point(-5.0, 0.0, -5.0) ));
        assert!( local_ray.direction.equals( vector(0.0, 0.0, 1.0) ));
    }

    #[test]
    fn computing_normal_on_translated_shape(){
        let mut s = shape();
        s.set_transform( translation(0.0, 1.0, 0.0) );

        let n = s.normal_at( point(0.0, 1.70711, -0.70711) );
        assert!( n.equals( vector(0.0, 0.70711, -0.70711) ));
    }

    #[test]
    fn computing_normal_on_transformed_shape(){
        let mut s = shape();
        s.set_transform( scaling(1.0, 0.5, 1.0).mult( &rotation_z( PI / 5.0 ) ));

        let n = s.normal_at( point(0.0, SQRT_2 / 2.0, -SQRT_2 / 2.0) );
        assert!( n.equals( vector(0.0, 0.97014, -0.24254) ));
    }

    #[test]
    fn a_shape_has_a_parent_attribute(){
        let s = shape();
        assert!(s.i.parent == None);
    }

    #[test]
    fn converting_point_from_world_to_object_space(){
        let mut g1 = group();
        g1.set_transform(rotation_y(PI/2.0));

        let mut g2 = group();
        g2.set_transform(scaling(2.0, 2.0, 2.0));

        let mut s = sphere();
        s.set_transform(translation(5.0, 0.0, 0.0));

        g2.add_child(Box::new(s.clone()));
        g1.add_child(Box::new(g2));

        let p = s.world_to_object(point(-2.0, 0.0, -10.0));
        assert!(p.equals(point(0.0, 0.0, -1.0)));
    }
}

File: python/shapes.py
# to run tests: python -m unittest -v shapes

import unittest
import math
import matrices
import materials
import tuples


class Shape:                                          # 'abstract' base class
    def __init__(self):
        self.transform = matrices.identity()
        self.material = materials.material()
        self.parent = None

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return (self.transform == other.transform and
                    self.material == other.material)
        else:
            return False

    def intersect(self, r):
        local_ray = r.transform(self.transform.inverse())
        return self.local_intersect(local_ray)

    def local_intersect(self, r):                     # override in child classes
        self.saved_ray = r                            # this implementation for test purposes only
    
    def set_transform(self, t):
        self.transform = t

    def normal_at(self, pt, i):                       # i is only used by select subclasses for uv data
        local_point = self.world_to_object(pt)
        local_normal = self.local_normal_at(local_point, i)
        return self.normal_to_world(local_normal)

    def local_normal_at(self, pt, i):                    # override in child classes
        return tuples.vector(pt.x, pt.y, pt.z)        # this implementation for test purposes only

    def world_to_object(self, pt):
        if self.parent:
            pt = self.parent.world_to_object(pt)

        return self.transform.inverse() * pt

    def normal_to_world(self, normal):
        normal = self.transform.inverse().transpose() * normal
        normal.w = 0
        normal = normal.normalize()

        if self.parent:
            normal = self.parent.normal_to_world(normal)

        return normal


def test_shape():
    return Shape()

import rays
import transformations
import spheres
import groups
import intersections

class ShapeTestCase(unittest.TestCase):
    def test_a_shapes_default_transformation(self):
        s = test_shape()

        self.assertEqual(s.transform, matrices.identity())

    def test_assigning_a_transformation(self):
        s = test_shape()

        s.set_transform(transformations.translation(2, 3, 4))

        self.assertEqual(s.transform, transformations.translation(2, 3, 4))

    def test_a_shapes_default_materials(self):
        s = test_shape()

        self.assertEqual(s.material, materials.material())

    def test_assigning_a_materials(self):
        s = test_shape()
        m = materials.material()
        m.ambient = 1

        s.material = m

        self.assertEqual(s.material, m)

    def test_intersecting_a_scaled_shape_with_a_ray(self):
        s = test_shape()
        s.set_transform(transformations.scaling(2, 2, 2))
        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))

        xs = s.intersect(r)

        self.assertEqual(s.saved_ray.origin, tuples.point(0, 0, -2.5))
        self.assertEqual(s.saved_ray.direction, tuples.vector(0, 0, 0.5))

    def test_intersecting_a_translated_shape_with_a_ray(self):
        s = test_shape()
        s.set_transform(transformations.translation(5, 0, 0))
        r = rays.ray(tuples.point(0, 0, -5), tuples.vector(0, 0, 1))

        xs = s.intersect(r)

        self.assertEqual(s.saved_ray.origin, tuples.point(-5, 0, -5))
        self.assertEqual(s.saved_ray.direction, tuples.vector(0, 0, 1))

    def test_computing_normal_on_translated_shape(self):
        s = test_shape()
        s.set_transform(transformations.translation(0, 1, 0))
        i = intersections.intersection(1, s)

        n = s.normal_at(tuples.point(0, 1.70711, -0.70711), i)

        self.assertEqual(n, tuples.vector(0, 0.70711, -0.70711))

    def test_computing_normal_on_transformed_shape(self):
        s = test_shape()
        m = transformations.scaling(1, 0.5, 1) * transformations.rotation_z(math.pi/5)
        s.set_transform(m)
        i = intersections.intersection(1, s)

        n = s.normal_at(tuples.point(0, math.sqrt(2)/2, -math.sqrt(2)/2), i)

        self.assertEqual(n, tuples.vector(0, 0.97014, -0.24254))

    def test_a_shape_has_a_parent_attribute(self):
        s = test_shape()

        self.assertEqual(s.parent, None)

    def test_converting_point_from_world_to_object_space(self):
        g1 = groups.group()
        g1.set_transform(transformations.rotation_y(math.pi/2))

        g2 = groups.group()
        g2.set_transform(transformations.scaling(2, 2, 2))
        g1.add_child(g2)

        s = spheres.sphere()
        s.set_transform(transformations.translation(5, 0, 0))
        g2.add_child(s)

        p = s.world_to_object(tuples.point(-2, 0, -10))

        self.assertEqual(p, tuples.point(0, 0, -1))

    def test_converting_a_normal_from_object_to_world_space(self):
        g1 = groups.group()
        g1.set_transform(transformations.rotation_y(math.pi/2))

        g2 = groups.group()
        g2.set_transform(transformations.scaling(1, 2, 3))
        g1.add_child(g2)

        s = spheres.sphere()
        s.set_transform(transformations.translation(5, 0, 0))
        g2.add_child(s)

        n = s.normal_to_world(tuples.vector(math.sqrt(3)/3, math.sqrt(3)/3, math.sqrt(3)/3))

        self.assertEqual(n, tuples.vector(0.28571, 0.42857, -0.85714))

    def test_finding_normal_on_a_child_object(self):
        g1 = groups.group()
        g1.set_transform(transformations.rotation_y(math.pi/2))

        g2 = groups.group()
        g2.set_transform(transformations.scaling(1, 2, 3))
        g1.add_child(g2)

        s = spheres.sphere()
        s.set_transform(transformations.translation(5, 0, 0))
        g2.add_child(s)

        i = intersections.intersection(1, s)

        n = s.normal_at(tuples.point(1.7321, 1.1547, -5.5774), i)

        self.assertEqual(n, tuples.vector(0.28570, 0.42854, -0.85716))


# ---------------------------------------------------------------------------
if __name__ == '__main__':
    unittest.main()
